pkg/splunk/enterprise/types.go

```diff
@@ -40,6 +48,61 @@ const (
 	SplunkMonitoringConsole InstanceType = "monitoring-console"
 )
 
+const (
+	// Max retries for each phase of the App install pipeline
+	pipelinePhaseMaxRetryCount = 3
+)
+
+// PipelineWorker represents execution context used to run an app pkg worker thread
+type PipelineWorker struct {
+	//  to the AppSource Spec entry
+	appSrcName string
+
+	// Reference to the App Framework Config
+	afwConfig *enterpriseApi.AppFrameworkSpec
+
+	// Reference to the App context from the CR status
+	appDeployInfo *enterpriseApi.AppDeploymentInfo
+
+	// Used for pod copy and install
+	targetPodName string
+
+	// runtime client
+	client *splcommon.ControllerClient
+
+	// cr meta object
+	cr splcommon.MetaObject
+
+	// statefulset to know replicaset details
+	sts *appsv1.StatefulSet
+
+	// isActive indicates if a worker is assigned
+	isActive bool
+
+	// waiter reference to inform the caller
+	waiter *sync.WaitGroup
+}
+
+// PipelinePhase represents one phase in the overall installation pipeline
+type PipelinePhase struct {
+	mutex        sync.Mutex
+	q            []*PipelineWorker
+	msgChannel   chan *PipelineWorker
+	workerWaiter sync.WaitGroup
+}
+
+// AppInstallPipeline defines the pipeline for the installation activity
+type AppInstallPipeline struct {
```

Comment:
Maybe the best place for all these data structures is `common_types.go`?

Reply:
We are using the common_types.go for all the data structures that are common to both the K8 API and the other places of the code. All these workers are short living, we better leave here. Probably, we can move all the AFW data structures to a separate file, in the future.